// @superdevhq/llm-integration/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  LLM: () => LLM,
  llm: () => llm
});
module.exports = __toCommonJS(index_exports);

// src/llm.ts
var import_meta = {};
var BASE_URL = (
  //@ts-expect-error VITE_BASE_URL is not defined in the environment
  import_meta.env?.VITE_BASE_URL ?? (typeof window === "undefined" ? process?.env?.VITE_BASE_URL : "")
);
var LLM = class {
  async generateText({
    prompt,
    schema,
    verifyJwt,
    supabaseClient
  }) {
    if (verifyJwt && !supabaseClient) {
      throw new Error(
        "Supabase client is required for JWT verification but is not available."
      );
    }
    let session = null;
    if (verifyJwt) {
      const {
        data: { session: currentSession }
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } = await supabaseClient.auth.getSession();
      if (!currentSession) {
        throw new Error(
          "User not authenticated, cannot verify JWT for LLM call. Please authenticate first."
        );
      }
      session = currentSession;
    }
    const response = await fetch(`${BASE_URL}/api/integrations/invoke-llm`, {
      method: "POST",
      body: JSON.stringify({
        prompt,
        ...schema ? { schema } : {}
      }),
      headers: {
        "Content-Type": "application/json",
        // Conditionally add Authorization header only if verifyJwt is true and session was fetched
        ...verifyJwt && session?.access_token ? {
          Authorization: `Bearer ${session?.access_token}`
        } : {}
      }
    });
    if (!response.ok) {
      const errorBody = await response.text();
      throw new Error(
        `LLM API request failed with status ${response.status}: ${errorBody}`
      );
    }
    const data = await response.json();
    if (schema) {
      return data;
    }
    return data.response;
  }
  async generateImage(prompt, options) {
    const response = await fetch(
      `${BASE_URL}/api/integrations/generate-image`,
      {
        method: "POST",
        body: JSON.stringify({
          prompt,
          ...options
        }),
        headers: {
          "Content-Type": "application/json"
        }
      }
    );
    if (!response.ok) {
      const errorBody = await response.text();
      throw new Error(
        `Image generation API request failed with status ${response.status}: ${errorBody}`
      );
    }
    const { data = { images: [] } } = await response.json();
    if (!data.images || !data.images.length) {
      throw new Error(
        "Invalid response format from image generation API: No images returned."
      );
    }
    return data.images[0].url;
  }
};

// src/index.ts
var llm = new LLM();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LLM,
  llm
});
