// @superdevhq/web-search/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  WebSearch: () => WebSearch,
  WebSearchError: () => WebSearchError,
  WebSearchOutputSchema: () => WebSearchOutputSchema,
  WebSearchParamsSchema: () => WebSearchParamsSchema,
  WebSearchResultSchema: () => WebSearchResultSchema,
  webSearch: () => webSearch
});
module.exports = __toCommonJS(index_exports);

// src/error.ts
var WebSearchError = class extends Error {
  statusCode;
  body;
  constructor(message, statusCode, body = "") {
    super(message);
    this.name = "WebSearchError";
    this.statusCode = statusCode;
    this.body = body;
  }
  /**
   * Returns true if the error is due to rate limiting
   */
  isRateLimited() {
    return this.statusCode === 429 || this.statusCode === 432 || this.statusCode === 433;
  }
  /**
   * Returns true if the error is due to authentication issues
   */
  isAuthError() {
    return this.statusCode === 401 || this.statusCode === 403;
  }
  /**
   * Returns true if the error is due to invalid input
   */
  isInputError() {
    return this.statusCode === 400;
  }
  /**
   * Returns true if the error is a server-side error
   */
  isServerError() {
    return this.statusCode >= 500;
  }
};

// src/web-search.ts
var import_meta = {};
var BASE_URL = (
  //@ts-expect-error VITE_BASE_URL is not defined in the environment
  import_meta.env?.VITE_BASE_URL ?? (typeof window === "undefined" ? process?.env?.VITE_BASE_URL : "")
);
var WebSearch = class {
  appId;
  apiKey;
  constructor(appId) {
    this.appId = appId;
  }
  async search(input) {
    const headers = {
      "Content-Type": "application/json"
    };
    headers["x-app-id"] = this.appId;
    let response;
    try {
      response = await fetch(`${BASE_URL}/api/integrations/web-search`, {
        method: "POST",
        headers,
        body: JSON.stringify(input)
      });
    } catch (error) {
      throw new WebSearchError(
        "Failed to connect to web search API",
        0,
        error instanceof Error ? error.message : String(error)
      );
    }
    const errorBody = !response.ok ? await response.text() : "";
    if (!response.ok) {
      throw new WebSearchError(
        `Web search API request failed with status ${response.status}`,
        response.status,
        errorBody
      );
    }
    let data;
    try {
      data = await response.json();
    } catch (error) {
      throw new WebSearchError(
        "Failed to parse API response",
        response.status,
        error instanceof Error ? error.message : String(error)
      );
    }
    if (data.error) {
      throw new WebSearchError(
        data.error,
        response.status,
        JSON.stringify(data)
      );
    }
    return data;
  }
};

// src/schema.ts
var import_zod = require("zod");
var WebSearchParamsSchema = import_zod.z.object({
  query: import_zod.z.string().min(1, "Search query is required"),
  max_results: import_zod.z.number().int().min(1).max(10).optional().default(5),
  include_images: import_zod.z.boolean().optional().default(false),
  include_answer: import_zod.z.boolean().optional().default(true),
  search_depth: import_zod.z.enum(["basic", "advanced"]).optional().default("basic")
});
var WebSearchResultSchema = import_zod.z.object({
  id: import_zod.z.string(),
  url: import_zod.z.string().url(),
  title: import_zod.z.string(),
  content: import_zod.z.string(),
  domain: import_zod.z.string(),
  score: import_zod.z.number().min(0).max(1)
});
var WebSearchOutputSchema = import_zod.z.object({
  query: import_zod.z.string(),
  results: import_zod.z.array(WebSearchResultSchema),
  answer: import_zod.z.string().optional()
});

// src/index.ts
var import_meta2 = {};
var webSearch = new WebSearch(
  //@ts-ignore
  import_meta2?.env?.VITE_APP_ID || ""
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  WebSearch,
  WebSearchError,
  WebSearchOutputSchema,
  WebSearchParamsSchema,
  WebSearchResultSchema,
  webSearch
});
