// @superdevhq/google-workspace/dist/index.js
// src/google-workspace.ts
var BASE_URL = (
  //@ts-expect-error VITE_BASE_URL is not defined in the environment
  import.meta.env.VITE_SUPERDEV_BASE_URL ?? "https://superdev.build/api"
);
var GoogleWorkspace = class {
  appId;
  constructor(appId) {
    this.appId = appId;
  }
  async getSheetValues(input) {
    const response = await fetch(`${BASE_URL}/integrations/google-workspace`, {
      method: "POST",
      body: JSON.stringify({
        action: "getSheetValues",
        spreadsheetId: input.spreadsheetId,
        range: input.range
      }),
      headers: {
        "Content-Type": "application/json",
        "x-app-id": this.appId
      }
    });
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    const data = await response.json();
    return data;
  }
  async updateSheetValues(input) {
    const response = await fetch(`${BASE_URL}/integrations/google-workspace`, {
      method: "POST",
      body: JSON.stringify({
        action: "updateSheetValues",
        spreadsheetId: input.spreadsheetId,
        range: input.range,
        values: input.values
      }),
      headers: {
        "Content-Type": "application/json",
        "x-app-id": this.appId
      }
    });
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    const data = await response.json();
    return data;
  }
  // async updateSheetValues(input: UpdateSheetValues.UpdateSheetValuesInput) {
  //   return updateSheetValues(input);
  // }
  async getSpreadsheetStructure(input) {
    const response = await fetch(`${BASE_URL}/integrations/google-workspace`, {
      method: "POST",
      body: JSON.stringify({
        action: "getSpreadsheetStructure",
        spreadsheetId: input.spreadsheetId
      }),
      headers: {
        "Content-Type": "application/json",
        "x-app-id": this.appId
      }
    });
    if (!response.ok) {
      const errorBody = await response.text();
      throw new Error(
        `Google Workspace API request failed with status ${response.status}: ${errorBody}`
      );
    }
    const data = await response.json();
    if (data.error) {
      throw new Error(data.error);
    }
    return data;
  }
  // async listDocuments(input: ListDocuments.ListDocumentsInput) {
  //   return listDocuments(input);
  // }
  // async getDocument(input: GetDocument.GetDocumentInput) {
  //   return getDocument(input);
  // }
  // async getDocumentAsMarkdown(
  //   input: GetDocumentAsMarkdown.GetDocumentAsMarkdownInput
  // ) {
  //   return getDocumentAsMarkdown(input);
  // }
  // async getDocumentAsMarkdown(
  //   input: GetDocumentAsMarkdown.GetDocumentAsMarkdownInput
  // ) {
  //   return getDocumentAsMarkdown(input);
  // }
};

// src/index.ts
var googleWorkspace = new GoogleWorkspace(
  //@ts-ignore
  import.meta?.env?.VITE_APP_ID || ""
);
export {
  GoogleWorkspace,
  googleWorkspace
};
